shared/domains/chelonia/internals.js

```diff
@@ -158,47 +162,220 @@ export default (sbp('sbp/selectors/register', {
       }
     }
   },
+  'chelonia/private/out/latestHash': function (contractID: string) {
+    return fetch(`${this.config.connectionURL}/latestHash/${contractID}`, {
+      cache: 'no-store'
+    }).then(handleFetchResult('text'))
+  },
+  // TODO: r.body is a stream.Transform, should we use a callback to process
+  //       the events one-by-one instead of converting to giant json object?
+  //       however, note if we do that they would be processed in reverse...
+  'chelonia/private/out/eventsSince': async function (contractID: string, since: string) {
+    const events = await fetch(`${this.config.connectionURL}/eventsSince/${contractID}/${since}`)
+      .then(handleFetchResult('json'))
+    if (Array.isArray(events)) {
+      return events.reverse().map(b64ToStr)
+    }
+  },
   'chelonia/private/in/processMessage': async function (message: GIMessage, state: Object) {
     const [opT, opV] = message.op()
     const hash = message.hash()
     const contractID = message.contractID()
     const manifestHash = message.manifest()
     const config = this.config
-    if (!state._vm) state._vm = {}
+    const signature = message.signature()
+    const signedPayload = message.signedPayload()
+    const env = this.env
+    const self = this
+    if (!state._vm) state._vm = Object.create(null)
     const opFns: { [GIOpType]: (any) => void } = {
       [GIMessage.OP_CONTRACT] (v: GIOpContract) {
-        // TODO: shouldn't each contract have its own set of authorized keys?
-        if (!state._vm.authorizedKeys) state._vm.authorizedKeys = []
-        // TODO: we probably want to be pushing the de-JSON-ified key here
-        state._vm.authorizedKeys.push({ key: v.keyJSON, context: 'owner' })
+        const keys = { ...env.additionalKeys, ...state._volatile?.keys }
+        state._vm.authorizedKeys = keysToMap(v.keys)
+
+        for (const key of v.keys) {
+          if (key.meta?.private) {
+            if (key.id && key.meta.private.keyId in keys && key.meta.private.content) {
+              if (!state._volatile) state._volatile = Object.create(null)
+              if (!state._volatile.keys) state._volatile.keys = Object.create(null)
+              try {
+                state._volatile.keys[key.id] = decrypt(keys[key.meta.private.keyId], key.meta.private.content)
+              } catch (e) {
+                console.error('Decryption error', e)
```

Comment:
Should we allow it to be thrown - or rethrow here? In other places in this selector we seem to throw exceptions instead of ignoring them.

Reply:
This allows the code to proceed if the encrypted data are invalid for some reason. Should we not ignore this error? Otherwise, it seems like a 'bad' key would result in a contract being permanently invalid

Reply:
For now then, could you please add a `// NOTE: ...` comment here explaining why it's being ignored?

Reply:
Note added in my PR